src/domain/datasets/services/src/use_cases/update_version_file_use_case_impl.rs

```diff
@@ -0,0 +1,227 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::sync::Arc;
+
+use dill::{component, interface};
+use file_utils::MediaType;
+use internal_error::{ErrorIntoInternal, InternalError, ResultIntoInternal};
+use kamu_core::{
+    DatasetRegistry,
+    FileUploadLimitConfig,
+    GetDataOptions,
+    PushIngestDataError,
+    PushIngestDataUseCase,
+    PushIngestError,
+    QueryService,
+};
+use kamu_datasets::{
+    ContentArgs,
+    ExtraDataFields,
+    FileVersion,
+    UpdateVersionFileResult,
+    UpdateVersionFileUseCase,
+    UpdateVersionFileUseCaseError,
+    UploadTooLargeError,
+    VersionedFileEntity,
+};
+use serde_json::Value;
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[component]
+#[interface(dyn UpdateVersionFileUseCase)]
+pub struct UpdateVersionFileUseCaseImpl {
+    dataset_registry: Arc<dyn DatasetRegistry>,
+    upload_config: Arc<FileUploadLimitConfig>,
+    push_ingest_data_use_case: Arc<dyn PushIngestDataUseCase>,
+    query_svc: Arc<dyn QueryService>,
+}
+
+impl UpdateVersionFileUseCaseImpl {
+    async fn get_latest_version(
+        &self,
+        dataset_ref: &odf::DatasetRef,
+    ) -> Result<(FileVersion, odf::Multihash), InternalError> {
+        // TODO: Consider retractions / corrections
+        let query_res = self
+            .query_svc
+            .tail(dataset_ref, 0, 1, GetDataOptions::default())
+            .await
+            .int_err()?;
```

Comment:
I think that needs to be rather an SQL query, if we have to consider corrections and retractions